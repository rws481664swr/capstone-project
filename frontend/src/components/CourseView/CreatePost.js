import {useState} from "react";

const CreatePost=({course})=>{
    const [text,setText]=useState('')
    const submit =(e)=>{
        e.preventDefault()
        //TODO POST TO COURSE
    }

    return <form onSubmit={submit}><input onChange={e=>setText(e.target.value)}
                        value={text}
                        type="text"/></form>
}
