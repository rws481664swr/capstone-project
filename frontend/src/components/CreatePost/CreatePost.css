.CreatePost{
    width:70%;
    margin :auto;
}
button.Button.CreatePost{
    width:70%;
    margin:auto;
}
form.CreatePost{
    /*display: flex;*/
    /*justify-content: center;*/
}
