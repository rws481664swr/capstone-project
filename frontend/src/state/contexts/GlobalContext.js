import {createContext, useContext} from "react";
import jwtDecode from "jwt-decode";
import useLocalStorageState from "../../hooks/useLocalStorageState";

const value = {
    username: undefined,
    role: undefined,
    _id: undefined,
    timestamp: undefined,
    token:null,
    setToken:()=>{}

}
const GlobalContext = createContext(value)
export default GlobalContext

export const GlobalContextProvider = ({children}) => {
    const [context,setContext] =
        useLocalStorageState( 'ctx',value||{})
    const setToken= (token)=>{
        if (token){
            const {

                username,
                role,
                _id,
                timestamp
            } = jwtDecode(token)
            setContext({username, role, _id, timestamp, token})
        }else{
            setContext({})
        }
    }
    const logout= ()=> {
        setToken(null)
    }
    return <GlobalContext.Provider value={{...context,setToken,logout}}>
        {children}
    </GlobalContext.Provider>
}
export const useGlobalContext = () =>     useContext(GlobalContext)

