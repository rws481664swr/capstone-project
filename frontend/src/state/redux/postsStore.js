import {configureStore} from "@reduxjs/toolkit";
import postsReducer from "./postsReducer";

const postsStore =
    configureStore({
        reducer: postsReducer
    })
export default postsStore