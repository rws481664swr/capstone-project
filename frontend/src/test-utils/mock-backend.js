import {BASE_URL} from "../config";

export const extractKeysAsList= (obj)=>
    Object.keys(obj).map(key=>obj[key])
        .reduce((acc, val)=>[...acc,val],[])
export const MOCK_TOKEN = {current: 'TOKEN'}

export const spliceUrl = (url) => {
    const arr = BASE_URL.split('')
    const toReturn = url.split('')
    while (arr.length){
        arr.shift()
        toReturn.shift()
    }
    return toReturn.join('')

}