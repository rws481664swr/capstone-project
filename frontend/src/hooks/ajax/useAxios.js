import axios from "axios";
import {BASE_URL} from "../../config";
import {useGlobalContext} from "../../state/contexts/GlobalContext";
import {useCallback, useEffect, useMemo} from "react";

const dateIsLessThanOneMinuteOld = (date) => {
    const now = new Date()
    const then = new Date(date)
    const diff = (now.getTime() - then.getTime()) / 1000
    return diff < 60
}

const NETWORK_ERROR = 'ERR_NETWORK'
const useAxios = () => {

    const {token:tokenRef,timestamp} = useGlobalContext()

    useEffect(() => {

        (async () => {
            if(dateIsLessThanOneMinuteOld(timestamp))return //do not need to refresh token
            const {data: {token: newToken}} = await axios.get(`${BASE_URL}/auth/token`,
                {headers: {authorization: `Bearer ${tokenRef.current}`}})
            tokenRef.current= newToken

        })()
    }, [ ]) // eslint-disable-line react-hooks/exhaustive-deps


    const config = useCallback((query = {}) => ({
        headers: {authorization: `Bearer ${tokenRef.current}`, 'Access-Control-Allow-Origin': '*'},
        params: query,
    }),[tokenRef])


   return useMemo (
       ()=> {

           async function get(resource, id = '', query = {}) {
               try {

                   const {data} =    await axios.get(`${BASE_URL}/${resource}/${id}`, config(query))
                   return data;

               }catch (e) {
                   if (e.code===NETWORK_ERROR){

                       return alert(`Site is down.`)
                   }
                   throw e
               }
           }

           async function post(resource, body) {
               try {
                   const {data} = await axios.post(`${BASE_URL}/${resource}`, body, config())
                   return data
               }catch (e) {
                   if (e.code===NETWORK_ERROR){
                       return alert(`Site is down.`)
                   }
                   throw e
               }
           }

           async function put(resource, id = '', body) {
               try {
                    const {data} = await axios.put(`${BASE_URL}/${resource}`, body, config())
                   return data
               }catch (e) {
                   if (e.code===NETWORK_ERROR){
                       return alert(`Site is down.`)
                   }
                   throw e
               }
           }

           async function delete_(resource, id = '') {
               try {
                   const {data} = await axios.delete(`${BASE_URL}/${resource}/${id}`, config())
                   return data;
               }catch (e) {
                 console.error(e)
                   if (e.code===NETWORK_ERROR){
                       return alert(`Site is down.`)
                   }
                   throw e
               }
           }

           return {post, put, delete: delete_, get}
       },[config])
}
export default useAxios