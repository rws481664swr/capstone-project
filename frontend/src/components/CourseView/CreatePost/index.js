import useForm from "../../../hooks/useForm";
import useFlash from "../../../hooks/useFlash";
import {useGlobalContext} from "../../../state/contexts/GlobalContext";
import useAxios from "../../../api";
import {ADD} from "../../../state/actions/posts";
import {useDispatch} from "react-redux";
import Modal from "../../General/Modal";
import {useCallback} from "react";


function useSubmitPost(course, form, clear, hide) {
    const dispatch = useDispatch()
    const {post:postRequest} = useAxios()
    const {_id: user, username} = useGlobalContext()
    const [msg, flash, setCss] = useFlash()
    const submit = useCallback(
    async (e) => {
        e.preventDefault()
        const danger = () => setCss('text-danger')
        const post = {...form, course: course._id, user, username}
        console.log('post', post)
        if (!post.content.trim() && !post.title.trim()) {
            danger()
            return flash('Title and content are required')
        }
        if (!post.title.trim()) {
            danger()
            return flash('Title is required')
        }
        if (!post.content.trim()) {
            danger()
            return flash('Content is required')
        }
        try{
            setCss('text-success')
            const payload = await postRequest('posts', post)
            dispatch({type: ADD, payload})
            clear()
            hide()
        }catch ({response:{data:{message:err}}}) {
            console.error(err)
            setCss('text-danger')
            return flash(err)
        }
    }
    , [hide,clear,dispatch,form,flash,setCss,username,user, course._id]) // eslint-disable-line react-hooks/exhaustive-deps
    return [submit, msg];
}

const CreatePost = ({course, hide}) => {
    const [form, onChange, clear] = useForm({title: '', content: ''})
    const [submit, msg] = useSubmitPost(course, form, clear, hide);
    return <>
        <Modal visible={true} hide={hide}>
            <form className={'CreatePost '} onSubmit={submit}>
                {course._id}
                <div className={'CreatePost'}> {msg}</div>
                <label className={'CreatePost '} htmlFor="post-title">Post Title</label>
                <input id="post-title" className={'CreatePost '}
                       onChange={onChange} value={form.title}
                       name={'title'}
                       type="text"/>

                <label className={'CreatePost '} htmlFor="post-content">Post Content</label>
                <textarea id={"post-content"} className={'CreatePost '}
                          name="content" cols="30" rows="10" value={form.content}
                          onChange={onChange}></textarea>

                <button className={'CreatePost '} type="submit">Post</button>
            </form>
        </Modal>
    </>
}
export default CreatePost