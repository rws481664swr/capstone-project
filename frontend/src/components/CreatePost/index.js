import useForm from "../../hooks/form/useForm";
import useFlash from "../../hooks/form/useFlash";
import {useGlobalContext} from "../../state/contexts/GlobalContext";
import useAxios from "../../hooks/ajax/useAxios";
import {ADD} from "../../state/actions/actions";
import {useDispatch} from "react-redux";
import Modal from "../General/Modal/Modal";
import {useCallback} from "react";
import './CreatePost.css'
import LabeledInput, {LabeledTextBox} from "../General/LabeledInput/LabeledInput";
import Button from "../General/Button/GenericButton/Button";

const formValidate = ({content, title}, flash, danger) => {
    content = content.trim()
    title = title.trim()
    let bad = false
    let msg = ''
    if (!title) {
        danger()
        msg = 'Title is required'
        bad = true
    }
    if (!content) {
        danger()
        msg = 'Content is required'
        bad = true
    }

    if (!content && !title) {
        danger()
        msg = 'Title and content are required'
        bad = true
    }

    if (bad) {
        danger()
        flash(msg)
    }
    return bad
}

function useSubmitPost(course, form, clear, hide) {
    const dispatch = useDispatch()
    const {post: postRequest} = useAxios()
    const {_id: user, username} = useGlobalContext()
    const [msg, flash, setCss] = useFlash()
    const submit = useCallback(
        async (e) => {
            e.preventDefault()
            const post = {...form, course: course._id, user, username}
            if (formValidate(post, flash, () => setCss('text-danger'))) return
            try {
                setCss('text-success')
                const payload = await postRequest('posts', post)
                dispatch({type: ADD, payload})
                clear()
                hide()
            } catch ({response: {data: {message: err}}}) {
                console.error(err)
                setCss('text-danger')
                return flash(err)
            }
        }
        , [hide, clear, dispatch, form, flash, setCss, username, user, course._id, postRequest])
    return [submit, msg];
}

const CreatePost = ({course, hide}) => {
    const [form, onChange, clear] = useForm({title: '', content: ''})
    const [submit, msg] = useSubmitPost(course, form, clear, hide);
    return <>
        <Modal visible={true} hide={hide}>
            <div>
                <form className={'CreatePost '} onSubmit={submit}>
                    <div className={'CreatePost'}> {msg}</div>
                    <LabeledInput
                        name={'title'}
                        onChange={(e) => {
                            onChange(e)
                        }}
                        label={'Post Title'}
                        id={'input-title'}
                        className={'CreatePost'}
                        value={form.title}
                    />
                    <LabeledTextBox
                        name={'content'}
                        value={form.content}
                        onChange={onChange}
                        label={'Post Content'}
                        id={'input-content'}
                        className={'CreatePost'}/>
                    {/*<label className={'CreatePost '} htmlFor="post-title">Post Title</label>*/}
                    {/*<input id="post-title" className={'CreatePost '}*/}
                    {/*       onChange={onChange} value={form.title}*/}
                    {/*       name={'title'}*/}
                    {/*       type="text"/>*/}

                    {/*<label className={'CreatePost '} htmlFor="post-content">Post Content</label>*/}
                    {/*<textarea id={"post-content"} className={'CreatePost '}*/}
                    {/*          name="content" cols="30" rows="10" value={form.content}*/}
                    {/*          onChange={onChange}></textarea>*/}

                    <Button className={'CreatePost '} type="submit">Post</Button>
                </form>
            </div>
        </Modal>
    </>
}
export default CreatePost