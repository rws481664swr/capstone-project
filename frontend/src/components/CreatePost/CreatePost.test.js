import CreatePost from "./index";
import {fireEvent, getByTestId, render, waitFor} from "@testing-library/react";
import {Provider} from "react-redux";
import getData from "../../mock_data/data";
import GlobalContext from "../../state/contexts/GlobalContext";
import postsStore from "../../state/redux/postsStore";
import {MOCK_TOKEN} from "../../../mock-backend";
import axios from "axios";
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect'
import '../../hooks/useAxios'
jest.mock('axios')
const {
    coursesList: [{_id: TEST_COURSE_ID}],
    usersList: [{_id, username}]
} = getData()
const hide = jest.fn()
const token = MOCK_TOKEN
it('should render', () => {
axios.get.mockImplementation(()=>Promise.resolve(
    {data: {token: MOCK_TOKEN}}
))

    render(
        <GlobalContext.Provider value={{_id, username,token}}>
            <Provider store={postsStore}>
                <CreatePost course={{_id: TEST_COURSE_ID}}
                            hide={hide}/>
            </Provider>
        </GlobalContext.Provider>
    )
})
it('should match snapshot', () => {
axios.get.mockImplementation(()=>Promise.resolve(
    {data: {token: MOCK_TOKEN}}
))
    const {asFragment}= render(
        <GlobalContext.Provider value={{_id, username,token}}>
            <Provider store={postsStore}>
                <CreatePost course={{_id: TEST_COURSE_ID}}
                            hide={hide}/>
            </Provider>
        </GlobalContext.Provider>
    )
    expect(asFragment()).toMatchSnapshot()
})
it('should fill out the form', async () => {
  axios.get.mockImplementation(()=>Promise.resolve(
      {data: {token: MOCK_TOKEN}}
  ))
    axios.post.mockImplementation ((url,body)=> {
        // console.log(url,body)
        console.log("MOCK POST", url, body)
       expect(body).toEqual({
           user: '646e33a759eab5cd39bbb8a7',
           username: 'admin',
           title: 'TEST_TITLE',
           content: 'TEST_CONTENT',
           course: TEST_COURSE_ID
       } )

        return  Promise.resolve( { data :{...body,_id:'TEST_ID'}})
    } )
   const {getByText,getByLabelText}= render(
        <GlobalContext.Provider value={{_id, username,token}}>
            <Provider store={postsStore}>
                <CreatePost course={{_id: TEST_COURSE_ID}}
                            hide={()=>{}}/>
            </Provider>
        </GlobalContext.Provider>
    )
    expect(postsStore.getState()).toEqual([])
    const title= getByLabelText('Post Title')

    const content= getByLabelText('Post Content')
    const button= getByText('Post')
    fireEvent.change(title,{target: {value:'TEST_TITLE'}})
    expect(title.value).toBe('TEST_TITLE')
    fireEvent.change(content,{target: {value:'TEST_CONTENT'}})
    expect(content.value).toBe('TEST_CONTENT')

    fireEvent.click(button)
    await waitFor(()=>{
         expect(axios.post).toHaveBeenCalledTimes(1)
        expect(title.value).toBe('')
        expect(content.value).toBe('')
        expect(postsStore.getState()).toEqual([{
            _id: 'TEST_ID',
            user: '646e33a759eab5cd39bbb8a7',
            username: 'admin',
            title: 'TEST_TITLE',
            content: 'TEST_CONTENT',
            course: TEST_COURSE_ID
        }])

    })
})