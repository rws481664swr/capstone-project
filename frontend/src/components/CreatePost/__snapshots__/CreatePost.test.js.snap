// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sr-modal"
    style="z-index: 1;"
  >
    <div
      class=" sr-modal_content"
      style="z-index: 1;"
    >
      <div>
        <form
          class="CreatePost "
        >
          <div
            class="CreatePost"
          >
             
          </div>
          <div
            class="LabeledInput-div CreatePost"
          >
            <label
              class="LabeledInput-input-label "
              for="input-title"
            >
              Post Title
            </label>
            <input
              class="LabeledInput-input-text "
              id="input-title"
              name="title"
              type="text"
              value=""
            />
          </div>
          <div
            class="LabeledInput-div CreatePost"
          >
            <label
              class="LabeledInput-label "
              for="input-content"
            >
              Post Content
            </label>
            <textarea
              class="LabeledInput-textarea "
              id="input-content"
              name="content"
            />
          </div>
          <button
            class="CreatePost  Button"
            type="submit"
          >
            Post
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
